File: src/features/scenarios/services/scenarioService.ts
import { mockAdapter } from '@/lib/api/mockAdapter'
import { apiClient } from '@/lib/api/apiClient'
import { endpoints } from '@/lib/api/endpoints'
import { shouldUseApi } from '@/lib/api/config'
import { STORAGE_KEYS } from '@/lib/constants'
import { readStorage, writeStorage } from '@/lib/storage/localStorage'
import { uid } from '@/lib/utils'
import type { SimulationResult } from '@/types/simulation'
import type {
  Escenario,
  EventoEscenario,
  PasoRuta,
  ResultadoEscenario,
  RutaAlternativa,
  ScenarioSnapshot,
  Workload,
} from '@/types/scenario'

const readAll = () => ({
  scenarios: readStorage<Escenario[]>(STORAGE_KEYS.scenarios, []),
  events: readStorage<EventoEscenario[]>(STORAGE_KEYS.scenarioEvents, []),
  results: readStorage<ResultadoEscenario[]>(STORAGE_KEYS.scenarioResults, []),
  routes: readStorage<RutaAlternativa[]>(STORAGE_KEYS.routes, []),
  steps: readStorage<PasoRuta[]>(STORAGE_KEYS.routeSteps, []),
})

const toSnapshot = (scenario: Escenario): ScenarioSnapshot => {
  const { events, results, routes, steps } = readAll()
  const scenarioEvents = events.filter((item) => item.escenarioId === scenario.id)
  const scenarioResults = results.filter((item) => item.escenarioId === scenario.id)
  const scenarioRoutes = routes.filter((item) => item.escenarioId === scenario.id)
  const routeIds = new Set(scenarioRoutes.map((route) => route.id))
  const scenarioSteps = steps.filter((item) => routeIds.has(item.rutaId))
  const bestRoute = scenarioRoutes.sort((a, b) => a.orden - b.orden)[0]
  const blocked = scenarioResults.filter(
    (item) => item.estadoSimulado === 'bloqueada',
  ).length
  const workload: Workload = bestRoute?.cargaTrabajo ?? 'media'
  return {
    escenario: scenario,
    eventos: scenarioEvents,
    resultados: scenarioResults,
    rutas: scenarioRoutes,
    pasosRuta: scenarioSteps,
    resumen: {
      materiasBloqueadas: blocked,
      creditosDisponibles: Math.max(12, 24 - blocked * 2),
      semestreEstimadoGraduacion: bestRoute?.semestreEstimadoGraduacion ?? 9,
      creditosPromedioSemestre:
        workload === 'alta' ? 22 : workload === 'media' ? 18 : 14,
      promedioProyectado: workload === 'alta' ? 3.95 : workload === 'media' ? 4.08 : 4.18,
      cargaTrabajo: workload,
    },
  }
}

export const scenarioService = {
  getScenarios(studentId: string) {
    if (shouldUseApi()) {
      return apiClient.get<ScenarioSnapshot[]>(
        endpoints.studentScenarios(studentId),
      )
    }
    return mockAdapter(() =>
      readAll()
        .scenarios.filter((scenario) => scenario.estudianteId === studentId)
        .map(toSnapshot),
    )
  },
  saveScenario(
    studentId: string,
    nombre: string,
    descripcion: string,
    result: SimulationResult,
  ) {
    if (shouldUseApi()) {
      return apiClient.post<
        {
          studentId: string
          nombre: string
          descripcion: string
          simulacion: SimulationResult
        },
        ScenarioSnapshot
      >(endpoints.scenarios, {
        studentId,
        nombre,
        descripcion,
        simulacion: result,
      })
    }
    return mockAdapter(() => {
      const store = readAll()
      if (
        store.scenarios.filter((item) => item.estudianteId === studentId).length >= 20
      ) {
        throw new Error('Solo puedes guardar hasta veinte escenarios en esta demo')
      }
      const now = new Date().toISOString()
      const scenarioId = uid('scenario')
      const scenario: Escenario = {
        id: scenarioId,
        estudianteId: studentId,
        nombre,
        descripcion,
        creadoEn: now,
        actualizadoEn: now,
      }
      const event: EventoEscenario = {
        id: uid('event'),
        escenarioId: scenarioId,
        materiaId: result.evento.materiaId,
        tipoEvento: result.evento.tipoEvento,
      }
      const resultados: ResultadoEscenario[] = result.historialSimulado.map((item) => ({
        id: uid('result'),
        escenarioId: scenarioId,
        materiaId: item.materiaId,
        estadoSimulado: item.estado,
      }))
      const rutas = result.rutas.map((route, index) => ({
        ...route,
        id: uid('route'),
        escenarioId: scenarioId,
        orden: index + 1,
      }))
      const courseChain = [result.evento.materiaId, ...result.materiasBloqueadas]
      const routeSteps: PasoRuta[] = rutas.flatMap((route) =>
        courseChain.map((materiaId, index) => ({
          id: uid('step'),
          rutaId: route.id,
          materiaId,
          semestreSugerido:
            route.semestreEstimadoGraduacion -
            route.duracionEstimada +
            1 +
            Math.min(index, route.duracionEstimada - 1),
          orden: index + 1,
        })),
      )

      writeStorage(STORAGE_KEYS.scenarios, [...store.scenarios, scenario])
      writeStorage(STORAGE_KEYS.scenarioEvents, [...store.events, event])
      writeStorage(STORAGE_KEYS.scenarioResults, [...store.results, ...resultados])
      writeStorage(STORAGE_KEYS.routes, [...store.routes, ...rutas])
      writeStorage(STORAGE_KEYS.routeSteps, [...store.steps, ...routeSteps])
      return toSnapshot(scenario)
    })
  },
  compareScenarios(scenarioAId: string, scenarioBId: string) {
    if (shouldUseApi()) {
      return apiClient.post<
        { scenarioAId: string; scenarioBId: string },
        { a: ScenarioSnapshot; b: ScenarioSnapshot }
      >(endpoints.compareScenarios, {
        scenarioAId,
        scenarioBId,
      })
    }
    return mockAdapter(() => {
      const { scenarios } = readAll()
      const a = scenarios.find((item) => item.id === scenarioAId)
      const b = scenarios.find((item) => item.id === scenarioBId)
      if (!a || !b) throw new Error('Selecciona dos escenarios válidos')
      return {
        a: toSnapshot(a),
        b: toSnapshot(b),
      }
    })
  },
}

File: backend/app/services/scenario_service.py
from datetime import datetime, timezone
from uuid import uuid4

from fastapi import HTTPException, status
from sqlalchemy import func, select
from sqlalchemy.orm import Session

from app.models.entities import RouteStep, Scenario, ScenarioEvent, ScenarioResult, SuggestedRoute
from app.models.enums import CourseStatus, Workload
from app.schemas.simulation import (
    RouteResponse,
    RouteStepResponse,
    ScenarioEventRecordResponse,
    ScenarioSnapshotResponse,
    ScenarioSummaryResponse,
)


def uid(prefix: str) -> str:
    return f"{prefix}_{uuid4().hex[:12]}"


def scenario_snapshot(db: Session, scenario: Scenario) -> ScenarioSnapshotResponse:
    events = list(
        db.scalars(select(ScenarioEvent).where(ScenarioEvent.escenario_id == scenario.id)),
    )
    results = list(
        db.scalars(select(ScenarioResult).where(ScenarioResult.escenario_id == scenario.id)),
    )
    routes = list(
        db.scalars(
            select(SuggestedRoute)
            .where(SuggestedRoute.escenario_id == scenario.id)
            .order_by(SuggestedRoute.orden),
        ),
    )
    route_ids = [route.id for route in routes]
    steps = (
        list(db.scalars(select(RouteStep).where(RouteStep.ruta_id.in_(route_ids))))
        if route_ids
        else []
    )
    best_route = routes[0] if routes else None
    blocked = sum(1 for item in results if item.estado_simulado == CourseStatus.BLOQUEADA)
    workload = best_route.carga_trabajo if best_route else Workload.MEDIA
    return ScenarioSnapshotResponse(
        escenario=scenario,
        eventos=[ScenarioEventRecordResponse.model_validate(item) for item in events],
        resultados=results,
        rutas=[RouteResponse.model_validate(item) for item in routes],
        pasos_ruta=[RouteStepResponse.model_validate(item) for item in steps],
        resumen=ScenarioSummaryResponse(
            materias_bloqueadas=blocked,
            creditos_disponibles=max(12, 24 - blocked * 2),
            semestre_estimado_graduacion=(
                best_route.semestre_estimado_graduacion if best_route else 9
            ),
            creditos_promedio_semestre=(
                22 if workload == Workload.ALTA else 18 if workload == Workload.MEDIA else 14
            ),
            promedio_proyectado=(
                3.95 if workload == Workload.ALTA else 4.08 if workload == Workload.MEDIA else 4.18
            ),
            carga_trabajo=workload,
        ),
    )


def save_scenario(
    db: Session,
    student_id: str,
    nombre: str,
    descripcion: str,
    simulation,
) -> ScenarioSnapshotResponse:
    count = db.scalar(
        select(func.count()).select_from(Scenario).where(Scenario.estudiante_id == student_id),
    )
    if count >= 20:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_CONTENT,
            detail="Solo puedes guardar hasta veinte escenarios en esta demo",
        )

    now = datetime.now(timezone.utc)
    scenario = Scenario(
        id=uid("scenario"),
        estudiante_id=student_id,
        nombre=nombre,
        descripcion=descripcion,
        creado_en=now,
        actualizado_en=now,
    )
    db.add(scenario)
    db.flush()

    db.add(
        ScenarioEvent(
            id=uid("event"),
            escenario_id=scenario.id,
            materia_id=simulation.evento.materia_id,
            tipo_evento=simulation.evento.tipo_evento,
        ),
    )
    for history in simulation.historial_simulado:
        db.add(
            ScenarioResult(
                id=uid("result"),
                escenario_id=scenario.id,
                materia_id=history.materia_id,
                estado_simulado=history.estado,
            ),
        )

    route_map: dict[str, SuggestedRoute] = {}
    for index, route in enumerate(simulation.rutas, start=1):
        stored_route = SuggestedRoute(
            id=uid("route"),
            escenario_id=scenario.id,
            nombre=route.nombre,
            orden=index,
            semestre_estimado_graduacion=route.semestre_estimado_graduacion,
            duracion_estimada=route.duracion_estimada,
            dificultad=route.dificultad,
            carga_trabajo=route.carga_trabajo,
            descripcion=route.descripcion,
        )
        db.add(stored_route)
        route_map[route.id] = stored_route

    course_chain = [simulation.evento.materia_id, *simulation.materias_bloqueadas]
    for stored_route in route_map.values():
        for index, materia_id in enumerate(course_chain, start=1):
            db.add(
                RouteStep(
                    id=uid("step"),
                    ruta_id=stored_route.id,
                    materia_id=materia_id,
                    semestre_sugerido=(
                        stored_route.semestre_estimado_graduacion
                        - stored_route.duracion_estimada
                        + min(index, stored_route.duracion_estimada)
                    ),
                    orden=index,
                ),
            )

    db.commit()
    db.refresh(scenario)
    return scenario_snapshot(db, scenario)


def compare_scenarios(
    db: Session,
    scenario_a_id: str,
    scenario_b_id: str,
) -> tuple[ScenarioSnapshotResponse, ScenarioSnapshotResponse]:
    scenario_a = db.get(Scenario, scenario_a_id)
    scenario_b = db.get(Scenario, scenario_b_id)
    if not scenario_a or not scenario_b:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Selecciona dos escenarios válidos",
        )
    return scenario_snapshot(db, scenario_a), scenario_snapshot(db, scenario_b)
